src/Exthost/Msg.re

```diff
@@ -502,6 +502,144 @@ module Telemetry = {
+    switch (method) {
+    | "$registerSourceControl" =>
+      switch (args) {
+      | `List([`Int(handle), `String(id), `String(label), rootUri]) =>
+        let rootUri = Uri.of_yojson(rootUri) |> Stdlib.Result.to_option;
+        Ok(RegisterSourceControl({handle, id, label, rootUri}));
+
+      | _ => Error("Unexpected arguments for $registerSourceControl")
+      }
+
+    | "$unregisterSourceControl" =>
+      switch (args) {
+      | `List([`Int(handle)]) =>
+        Ok(UnregisterSourceControl({handle: handle}))
+
+      | _ => Error("Unexpected arguments for $unregisterSourceControl")
+      }
+
+    | "$updateSourceControl" =>
+      switch (args) {
+      | `List([`Int(handle), features]) =>
+        Yojson.Safe.Util.(
+          Ok(
+            UpdateSourceControl({
+              handle,
+              hasQuickDiffProvider:
+                features |> member("hasQuickDiffProvider") |> to_bool_option,
+              count: features |> member("count") |> to_int_option,
+              commitTemplate:
+                features |> member("commitTemplate") |> to_string_option,
+              acceptInputCommand:
+                features |> member("acceptInputCommand") |> SCM.Decode.command,
+            }),
+          )
+        )
+
+      | _ => Error("Unexpected arguments for $updateSourceControl")
+      }
+
+    | "$registerGroup" =>
+      switch (args) {
+      | `List([`Int(provider), `Int(handle), `String(id), `String(label)]) =>
+        Ok(RegisterSCMResourceGroup({provider, handle, id, label}))
+
+      | _ => Error("Unexpected arguments for $registerGroup")
+      }
+
+    | "$unregisterGroup" =>
+      switch (args) {
+      | `List([`Int(handle), `Int(provider)]) =>
+        Ok(UnregisterSCMResourceGroup({provider, handle}))
+
+      | _ => Error("Unexpected arguments for $unregisterGroup")
+      }
+
+    // TODO:
+    /*| "$spliceResourceStates" =>
+        switch (args) {
+        | `List([`Int(provider), `List(groupSplices)]) =>
+          List.map(
+            fun
+            | `List([`Int(group), `List(splices)]) =>
+              List.map(
+                splice =>
+                  switch (splice) {
+                  | `List([`Int(start), `Int(deleteCount), `List(additions)]) =>
+                    let additions = List.map(SCM.Decode.resource, additions);
+                    Ok(
+                      SpliceSCMResourceStates({
+                        provider,
+                        group,
+                        start,
+                        deleteCount,
+                        additions,
+                      }),
+                    );
+
+                  | _ => Error("spliceResourceStates: Unexpected json")
+                  },
+                splices,
+              )
+            | _ => Error("spliceResourceStates: Unexpected json"),
+            groupSplices,
+          )
+
+        | _ => Error("Unexpected arguments for $spliceResourceStates")
+      };
```

Comment:
Need to flatten / adjust these types so that we can have a single `Ok(...)` result from it - the previous implementation relied on making multiple dispatches with a nested `List.iter`